docs/philosophy.rst

```diff
@@ -66,13 +66,168 @@ See this example for reference::
+              return "denied, cannot borrow any more books"
+          else:
+              self.book_storage.update_books(
+                  update=BookStorage.UpdateBooksModel(
+                      book_ids=books, borrowed_to=customer_id
+                  )
+              )
+              return "success, books are lended to customer"
+
+Notice how the implementation of the use case does not depend on the
+concret implementation of the BookStorage but instead only on the
+protocol.  This allows ``BookStorageSqlImplementation`` to change
+independently from the ``BorrowBookUseCase`` and vice versa.  A side
+benefit is that we can easily change the implementation of
+``BookStorage`` to another one. This comes handy when writing test
+code. Here is an example for how one might implement a test for the
+use case without the need to create an SQL database::
+
+  # In our test case we don't want to use the sql implementation since
+  # it is "expensive" to create new tables and set up database
+  # schemes. Therefore we use a BookStorage implementation that just
+  # stores all the relevant information in a python dictionary.
+  class UseCaseTests(TestCase):
+      def setUp(self) -> None:
+          self.book_storage = BookStorageTestImpl()
+          self.use_case = BorrowBookUseCase(book_storage=self.book_storage)
+
+      def test_customer_cannot_borrow_books_if_they_already_borrowed_20(self) -> None:
+          customer_id = 42
+          for book_id in range(20):
+              self.book_storage.create_book(book_id=book_id, borrowed_to=customer_id)
+          response = self.use_case.borrow_book(customer_id, books=[1, 2, 3])
+          self.assertTrue(response.startswith("denied"))
+
+
+  # This class implements all the methods defined in the protocol
+  # BookStorage. Notice how we can define additional methods like the
+  # create_book method that is not declared in the protocol.
+  class BookStorageTestImpl:
+      def __init__(self) -> None:
+          # This dictionary stores book ids and the customer id for
+          # potential customers that borrowed the specific book.
+          self.books: Dict[int, Optional[int]] = dict()
+
+      def get_book_count_borrowed_by_customer(self, customer_id: int) -> int:
+          return sum(
+              1 for _ in filter(lambda _id: _id == customer_id, self.books.values())
+          )
+
+      def update_books(self, update: BookStorage.UpdateBooksModel) -> None:
+          if update.borrowed_to is None:
+              return
+          customer_id = update.book_ids
+          for book in update.book_ids:
+              self.books[book] = customer_id
+
+      def create_book(self, book_id: int, borrowed_to: Optional[int]) -> None:
+          self.books[book_id] = borrowed_to
+
+The protocols that define DB access in the *arbeitszeitapp* are
+located under ``arbeitszeit.repositories``. Note that while the above
+method is the standard for new implementations there is some "legacy"
```

Comment:
a suggestion, to make clearer, what the new standard is: "Not that while the above method, *using protocol*, is the standard for ..." 
?